module/Photo/src/Photo/Mapper/Photo.php

```diff
@@ -30,12 +30,79 @@ public function __construct(EntityManager $em)
     }
 
     /**
+     * Returns the next photo in the album to display
+     * 
+     * @param \Photo\Model\Photo $photo 
+     */
+    public function getNextPhoto($photo)
+    {
+        $qb = $this->em->createQueryBuilder();
+
+        $qb->select('a')
+                ->from('Photo\Model\Photo', 'a')
+                ->where('a.id > ?1 AND a.album = ?2');
+        $qb->setParameter(1, $photo->getId());
+        $qb->setParameter(2, $photo->getAlbum());
+        $qb->addOrderBy('a.id', 'ASC');
+        $qb->setMaxResults(1);
+        $res = $qb->getQuery()->getResult();
+        return empty($res) ? null : $res[0];
+    }
+
+    /**
+     * Returns the previous photo in the album to display
+     * 
+     * @param \Photo\Model\Photo $photo 
+     */
+    public function getPreviousPhoto($photo)
+    {
+        $qb = $this->em->createQueryBuilder();
+
+        $qb->select('a')
+                ->from('Photo\Model\Photo', 'a')
+                ->where('a.id < ?1 AND a.album = ?2');
+        $qb->setParameter(1, $photo->getId());
+        $qb->setParameter(2, $photo->getAlbum());
+        $qb->addOrderBy('a.id', 'DESC');
+        $qb->setMaxResults(1);
+        $res = $qb->getQuery()->getResult();
+        return empty($res) ? null : $res[0];
+    }
+
+    /**
+     * Returns a random photo from the specified album
+     * 
+     * @param Photo\Model\Album $album
+     * @return Photo\Model\Photo
+     */
+    public function getRandomPhoto($album)
+    {
+        $qb = $this->em->createQueryBuilder();
+
+        /**
+         * This is not exactly the right way to do this. There is no way to just
+         * get a random row, unless we use a raw sql query.
+         * The doctrine team refuses to implement any 
+         * random functionality. (http://www.doctrine-project.org/jira/browse/DDC-950)
+         * 
+         */
+        $qb->select('a')
+                ->from('Photo\Model\Photo', 'a')
+                ->where('a.album = ?1');
+        $qb->setParameter(1, $album);
+        $qb->setMaxResults(20);
+        $res = $qb->getQuery()->getResult();
+        return empty($res) ? null : $res[array_rand($res)];
```

Comment:
Is het niet een soort van snel als je het zo doet: voor elk album de count bepalen (heb je mss al) en dan een sql limit 1 offset rand()%cnt in PHP? Als je order by een index doet is dat toch snel? 

Groetjes,

Nicky

(verzonden vanaf m'n telefoon)

> On 1 jan. 2015, at 22:01, jszanto <notifications@github.com> wrote:
> 
> In module/Photo/src/Photo/Mapper/Photo.php:
> 
> > +        $qb = $this->em->createQueryBuilder();
> > +
> > +        /**
> > +         * This is not exactly the right way to do this. There is no way to just
> > +         * get a random row, unless we use a raw sql query.
> > +         * The doctrine team refuses to implement any 
> > +         * random functionality. (http://www.doctrine-project.org/jira/browse/DDC-950)
> > +         * 
> > +         */
> > +        $qb->select('a')
> > +                ->from('Photo\Model\Photo', 'a')
> > +                ->where('a.album = ?1');
> > +        $qb->setParameter(1, $album);
> > +        $qb->setMaxResults(20);
> > +        $res = $qb->getQuery()->getResult();
> > +        return empty($res) ? null : $res[array_rand($res)];
> Een ander idee is om dit idee voor random cover foto's even te schrappen. Bij het importeren van foto's zouden we een cover foto kunnen genereren die 12 miniaturen van random photos uit het album bevat. Voorbeeld: http://i.imgur.com/yrIUyxR.jpg
> —
> Reply to this email directly or view it on GitHub.
>